pysmt/optimization/optimsat.py

```diff
@@ -127,54 +128,52 @@ def optimize(self, goal, **kwargs):
         msat_obj = self._assert_msat_goal(goal)
 
         self.solve()
-        optres = self._msat_lib.msat_objective_result(self.msat_env(), msat_obj)
-        if optres == self._msat_lib.MSAT_OPT_UNKNOWN:
-            raise SolverReturnedUnknownResultError()
-        elif optres == self._msat_lib.MSAT_OPT_UNSAT:
-            return None
-        else:
-            unbounded = self._msat_lib.msat_objective_value_is_unbounded(self.msat_env(),
-                                                                  msat_obj,
-                                                                  self._msat_lib.MSAT_OPTIMUM)
-            if unbounded > 0:
-                raise PysmtUnboundedOptimizationError("The optimal value is unbounded")
 
-            is_strict = self._msat_lib.msat_objective_value_is_strict(self.msat_env(),
-                                                                      msat_obj,
-                                                                      self._msat_lib.MSAT_OPTIMUM)
-            if is_strict:
-                raise PysmtUnboundedOptimizationError("The optimal value is infinitesimal")
-
-            check = self._msat_lib.msat_load_objective_model(self.msat_env(), msat_obj)
-            if check != 0:
-                raise ValueError()
+        if self._check_unsat_unbound_infinitesimal(msat_obj):
 
             model = self.get_model()
-            value = self._msat_lib.msat_objective_value_term(self.msat_env(),
-                                                             msat_obj,
-                                                             self._msat_lib.MSAT_OPTIMUM)
+            if goal.is_maxsmt_goal():
+                cost = self._compute_max_smt_cost(model, goal)
+                return model, cost
+            value = self._msat_lib.msat_objective_value_term(
+                self.msat_env(),
+                msat_obj,
+                self._msat_lib.MSAT_OPTIMUM
+            )
             return model, self.converter.back(value)
 
+        return None
+
     @clear_pending_pop
     def pareto_optimize(self, goals):
+        self._check_pareto_lexicographic_goals(goals, "pareto")
         self._msat_lib.msat_set_opt_priority(self.msat_env(), "par")
 
+        msat_objs = {}
+
         for g in goals:
-            self._assert_msat_goal(g)
+            msat_objs[g] = self._assert_msat_goal(g)
 
         while self.solve():
-            model = self.get_model()
-            yield model, [model.get_value(goal.term()) for goal in goals]
+            if all(self._check_unsat_unbound_infinitesimal(msat_objs[g]) for g in goals):
```

Comment:
```suggestion
            if all(self._check_unsat_unbound_infinitesimal(v) for v in goals.values()):
```

Reply:
`goals` is not a `dict`, did you mean `msat_objs.values()`?

Reply:
Yes, sorry :smile: